### ISSUE TITLE
Annotation alignment and overlap issues

### ISSUE THREAD

**Issue by mdeleeuw1** [DIAGNijmegen/pathology-hooknet#18]
Dear Mart and other DIAG members,

I have been trying to use HookNet on my dataset that consists of .mrxs WSIs and of .geojson annotations, thank you for your work on pathological image segmentation! Your last update to the model has resolved many of my issues, so I would like to thank you for this. However, I am still encountering two issues, which I shall outline below.

1. When loading my .mrxs WSIs using the WholeSlideImage class and then plotting the slide, all purely white parts of the WSI are turned to black / their values deleted, creating a blocky, white edge around the parts of the WSI that are not white (i.e., the tissue and some glass edges). Also, many black pixels are displayed around the original WSI, creating a sort of black, asymmetric rim around the original WSI when plotted. Since the relative position of the annotation is based on the position of the WSI (I presume) and the WSI is not aligned due to the additionally plotted black rim I mentioned before, they are not correctly aligned with one another (when plotting both the WSI and the annotations).
2. It seems as though the 'ignore_overlap' argument of the 'WholeSlideAnnotation' class is supposed to deal with overlapping annotations if set to False. However, when setting this argument either to False or True, my annotations still overlap when plotted with the plot_annotations function. I tried to tweak the _set_overlapping_annotations method to resolve this, but without any success.

It could very well be that these issues are due to a lack of experience in Python, but I would really appreciate hearing back from you.

Thank you in advance!

Regards,

Mike

**Comment by martvanrijthoven** [comment]
Dear Mike,

Thanks again for your interest in HookNet. It is great to hear that the last update resolved many issues for you. I hope i can help with these last issues.

1.
I have not encountered this problem myself. Would it be possible to share such a .mrxs and geojson with me, such that i can investigate it further?

2.
ignore_overlap is used when sampling (to make sure that a patch is not sampled with a center point of a different class), but it is actually not used in plot_annotations. Could you tell me what you need for annotations with regarding overlap? Maybe there is some other functionality you can use.

Best wishes,
Mart

**Comment by mdeleeuw1** [comment]
Dear Mart,

Thank you for the fast response and for helping me out.

1. I will share the WSI and its annotation files via email. It should appear in your inbox in a few minutes.

2. That makes sense, thanks for clarifying. I have several annotations with overlapping regions, and I would like to 'subtract' the overlapping regions from a few annotation types, creating a set of annotations that do not overlap and border each other for all WSIs.

I hope this makes sense - let me know if it doesn't!

Kind regards,

Mike

**Comment by mdeleeuw1** [comment]
Dear Mart,

I found a method to remove the overlapping regions through QuPath, so you can ignore issue #2. Thanks!

Kind regards,

Mike

**Comment by martvanrijthoven** [comment]
Dear Mike, 

Great to hear that you found a way to deal with the overlapping annotations in qupath!

For the other issue, I found out that, at least for visualisation of low resolution, you can use pyvips backend and the black parts around the WSI will not appear (to install pyvips you can use: conda install --channel conda-forge pyvips). Although when extracting patches  at high resolution the black borders seem to be still there. So not sure what is going on but it has something to do with openslide. Let me know if the black pixels become a problem and we can discuss a fix for this.

For the offset,  I found out that there is some offset setting in the mrxs. I implemented a fix in: https://github.com/DIAGNijmegen/pathology-whole-slide-data/issues/44

I will also send you a notebook in which I demonstrate how to use the fix.

Best wishes
Mart

**Comment by mdeleeuw1** [comment]
Dear Mart,

Hope you are well.

I have found another issue with the way that I am using HookNet. I am trying to train the model, but there seems to be some error in the way the y_batches are generated. I am sending you an email with some images of the errors and my training configuration. Hopefully you can make some sense of it, because I have been stuck here for a few days. Also, the loss function is often returning 'nan' often as value, there seems to be a problem when a patch contains an unannotated region (not sure, though). This could be linked to the other issue I am experiencing, but it seems to me that they are not correlated.

I hope you can make some sense of this, and if not, thanks in advance anyways.

Kind regards,

Mike

**Comment by martvanrijthoven** [comment]
This should be fixed now by https://github.com/DIAGNijmegen/pathology-whole-slide-data/commit/0a20034f237ed154f0d74f75bbd15432c80ddcc3

Please feel free to open another issue or reopen this issues if you still run into problems

**Comment by yuling-luo** [comment]
Hi @martvanrijthoven 
I hope you have a happy new year!
I have implemented the hooknet and can successfully run it : )
However, I also have a question regarding to the overlap annotation.
In my project, I only defined two classes A, B in the config (**B is always within A region**). I set the number of class in hooknet as 3 to take background into consideration. 
My question is: after create_batch_iterator and pytorch dataloader, the label tensor has shape of [batch_size, 3, h, w]. But because my classes overlap with each other, then it becomes a multi-label multi class problem. I wonder what are your thoughts on approaching this type of problem using hooknet?

**Comment by martvanrijthoven** [comment]
Dear Yuling Luo,

Happy new year to you too! It is really great to hear that you can run hooknet successfully!

Here are my thoughts on your problem:
- Hooknet should work well with a multiclass problem
- if you dont have background annotated, please make sure that A,B are exhaustively annotated otherwise these might appear as background (because a region including A,B is then not annotated). If A,B are not  exhaustively annotated you can make explicit annotations for background and sample from those.
- if B is always within A, make sure that B has a higher label value in. The default settings in wholeslidedata makes sure that B will be correctly on top of A in the generated masks. If it is not possible to give B a higher label value than A, you can use another sorting mechanism to make it work, but please let me know if you need it and i wil provide an example.

Please let me know if you have any further question and I am happy to address them!

Best wishes,
Mart

**Comment by yuling-luo** [comment]
> Dear Yuling Luo,
> 
> Happy new year to you too! It is really great to hear that you can run hooknet successfully!
> 
> Here are my thoughts on your problem:
> 
> * Hooknet should work well with a multiclass problem
> * if you dont have background annotated, please make sure that A,B are exhaustively annotated otherwise these might appear as background (because a region including A,B is then not annotated). If A,B are not  exhaustively annotated you can make explicit annotations for background and sample from those.
> * if B is always within A, make sure that B has a higher label value in. The default settings in wholeslidedata makes sure that B will be correctly on top of A in the generated masks. If it is not possible to give B a higher label value than A, you can use another sorting mechanism to make it work, but please let me know if you need it and i wil provide an example.
> 
> Please let me know if you have any further question and I am happy to address them!
> 
> Best wishes, Mart

Hi Mart,

Thank you so much for your response!

1. The background is not annotated but A,B are exhaustively annotated. So i assume this should be no issue.
2. B should have higher label value than A. It's either zero B annotation in an image or it has higher label value than A.
3. I'm interested to know how wholeslidedata covert the label annotation to array? Especially in my case where two annotation overlap? In the labels section in user_config, I set it as {A:1, B:2}. After create_batch_iterator, I can see that y_batch is an array that contains (0,1,2) and with a shape of [batch size, channel, h, w]. However, when I changed the user config and only set {A:1}, the y_batch still contain (0,1,2). Why is that?

**Comment by martvanrijthoven** [comment]
Dear Yuling Luo,

Ok then everything should work and it is indeed strange that you get values 0,1,2 when only specifying A:1, this should not happen.
Would it be possible to share your config file? Lets continue this discussion in your new [issue](https://github.com/DIAGNijmegen/pathology-hooknet/issues/21)
